src/node_writer.rs

```diff
@@ -0,0 +1,242 @@
+            bytes_written: 0,
+            path,
+            factory,
+        })
+    }
+
+    fn bytes_written(&self) -> usize {
+        self.bytes_written
+    }
+}
+
+impl Write for DataWriter {
+    fn write(&mut self, buf: &[u8]) -> Result<usize> {
+        let res = self.inner.write(buf);
+        if let Ok(size) = res {
+            self.bytes_written += size;
+        }
+        res
+    }
+
+    fn flush(&mut self) -> Result<()> {
+        self.inner.flush()
+    }
+}
+
+impl Drop for DataWriter {
+    fn drop(&mut self) {
+        // If we did not write anything into this node, it should not exist.
+        if self.bytes_written == 0 {
+            // We are ignoring deletion errors here in case the file is already gone.
+            let _ = self.factory.remove(&self.path);
+        }
+
+        // TODO(hrapp): Add some sanity checks that we do not have nodes with ridiculously low
+        // amount of points laying around?
+    }
+}
+
+#[derive(Default)]
+pub struct DataWriterBuilder {
+    factories: FnvHashMap<String, Arc<dyn CreateRemoveFactory>>,
+}
+
+impl DataWriterBuilder {
+    pub fn new() -> Self {
+        Self {
+            factories: FnvHashMap::default(),
+        }
+    }
+
+    pub fn add(mut self, prefix: impl Into<String>, factory: Box<dyn CreateRemoveFactory>) -> Self {
+        self.factories.insert(prefix.into(), Arc::from(factory));
+        self
+    }
+
+    pub fn create(&self, path: impl Into<PathBuf>) -> Result<DataWriter> {
+        let path = path.into();
+        for (prefix, factory) in &self.factories {
+            if !path.starts_with(prefix) {
+                continue;
+            }
+            return DataWriter::new(path, Arc::clone(factory));
+        }
+        DataWriter::new(path, Arc::from(OnDiskCreateRemoveFactory {}))
+    }
+}
+
+pub struct NodeWriter {
+    xyz_writer: DataWriter,
+    rgb_writer: DataWriter,
+    intensity_writer: Option<DataWriter>,
+    position_rescaling: Option<PositionRescaling>,
+    stem: PathBuf,
+    data_writer_builder: Arc<DataWriterBuilder>,
+}
+
+impl NodeWriter {
+    pub fn new(
+        path: impl Into<PathBuf>,
+        data_writer_builder: Arc<DataWriterBuilder>,
+        position_rescaling: Option<PositionRescaling>,
+    ) -> Self {
+        let stem: PathBuf = path.into();
+        NodeWriter {
+            xyz_writer: data_writer_builder
+                .create(&stem.with_extension(NodeLayer::Position.extension()))
+                .unwrap(),
+            rgb_writer: data_writer_builder
+                .create(&stem.with_extension(NodeLayer::Color.extension()))
+                .unwrap(),
+            intensity_writer: None, // Will be created if needed on first point with intensities.
+            stem,
+            data_writer_builder,
+            position_rescaling,
+        }
+    }
+
+    pub fn write(&mut self, p: &Point) {
+        // TODO(feuerste): This is expensive. It would be preferable not to branch on every point.
```

Comment:
could this function be operating on PointData? (does not need to be implemented on this PR

Reply:
Yes, this could definitely go into another PR, if it makes sense...